From 1fd73074d576e3cde639106324b5c2f07fac92af Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:12:03 -0400
Subject: [powerpc] powernv: Fix VFIO support with PHB3

Message-id: <1410545655-205645-495-git-send-email-dzickus@redhat.com>
Patchwork-id: 94470
O-Subject: [RHEL7 PATCH 494/626] powernv: Fix VFIO support with PHB3
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 08607afba64cc7a14d22b8aa2837d00f1d877d83
Author: Thadeu Lima de Souza Cascardo <cascardo@linux.vnet.ibm.com>
Date:   Mon Dec 9 14:41:01 2013 -0200

    powernv: Fix VFIO support with PHB3

    I have recently found out that no iommu_groups could be found under
    /sys/ on a P8. That prevents PCI passthrough from working.

    During my investigation, I found out there seems to be a missing
    iommu_register_group for PHB3. The following patch seems to fix the
    problem. After applying it, I see iommu_groups under
    /sys/kernel/iommu_groups/, and can also bind vfio-pci to an adapter,
    which gives me a device at /dev/vfio/.

    Signed-off-by: Thadeu Lima de Souza Cascardo <cascardo@linux.vnet.ibm.com>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/platforms/powernv/pci-ioda.c b/arch/powerpc/platforms/powernv/pci-ioda.c
index 30888a4..207911e 100644
--- a/arch/powerpc/platforms/powernv/pci-ioda.c
+++ b/arch/powerpc/platforms/powernv/pci-ioda.c
@@ -775,6 +775,7 @@ static void pnv_pci_ioda2_setup_dma_pe(struct pnv_phb *phb,
   tbl->it_type |= (TCE_PCI_SWINV_CREATE | TCE_PCI_SWINV_FREE);
  }
  iommu_init_table(tbl, phb->hose->node);
+ iommu_register_group(tbl, pci_domain_nr(pe->pbus), pe->pe_number);
 
  if (pe->pdev)
   set_iommu_table_base_and_group(&pe->pdev->dev, tbl);
-- 
1.7.1